radiometric_normalization/time_stack.py

```diff
@@ -0,0 +1,214 @@
+            masked array of the image data
+    '''
+
+    no_images = len(all_images)
+    no_bands, _, _ = all_images[0].shape
+    all_bands = []
+    for band in range(no_bands):
+        one_band = \
+            [numpy.ma.masked_equal(all_images[image][band, :, :], nodata)
+             for image in range(no_images)]
+        all_bands.append(one_band)
+
+    return all_bands
+
+
+def _mean_with_uniform_weight(all_bands, output_nodata, output_datatype):
+    ''' Calculates the time stack from a list of the bands
+
+    Input:
+        all_bands (list of list of arrays): A list of each band,
+            each band consisting of a list of images each entry being a
+            masked array of the image data
+        output_nodata (number): The no data value for the output image
+        output_datatype (numpy datatype): Data type for the output image
+
+    Output:
+        output_bands (list of arrays of numbers): A list of arrays, each
+            representing one band of the time stack
+        mask (array of numbers): An array representing the weight of the
+            time stack
+    '''
+
+    logging.info('Time stack analysis is using: Mean with uniform weight.')
+
+    rows, cols = all_bands[0][0].shape
+    all_bands_mask = numpy.zeros((rows, cols))  # 0 for valid; 1 for nodata
+    output_bands = []
+    for band in all_bands:
+        masked_mean = numpy.ma.mean(band, axis=0)
+        band_mean = masked_mean.data
+        band_mask = masked_mean.mask
+        band_mean[numpy.nonzero(band_mask)] = output_nodata
+        all_bands_mask = numpy.logical_or(band_mask, all_bands_mask)
+        output_bands.append(band_mean.astype(output_datatype))
+
+    output_mask = numpy.logical_not(all_bands_mask).astype(output_datatype) * \
+        numpy.iinfo(output_datatype).max
+
+    return output_bands, output_mask
+
+
+def _write_out_bands(output_bands, mask,
+                     output_path, output_nodata):
+    ''' Writes out the time stack to a GeoTIFF
+
+    Input:
+        output_bands (list of arrays of numbers): A list of arrays, each
+            representing one band of the time stack
+        mask (array of numbers): An array representing the weight of the
+            time stack
+        output_path (str): The path to save the image at
+        output_nodata (number): The no data value for the output image
+    '''
+
+    no_bands = len(output_bands)
+    rows, cols = output_bands[0].shape
+
+    # Set up output file
+    options = ['ALPHA=YES']
+    datatype = gdal_array.NumericTypeCodeToGDALTypeCode(
+        output_bands[0].dtype.type)
+    gdal_ds = gdal.GetDriverByName('GTIFF').Create(
+        output_path, cols, rows, no_bands + 1, datatype,
+        options=options)
+
+    # Write output file data
+    for i in range(no_bands):
+        gdal_array.BandWriteArray(
+            gdal_ds.GetRasterBand(i + 1),
+            output_bands[i])
+    gdal_ds.GetRasterBand(1).SetNoDataValue(output_nodata)
```

Comment:
I believe the nodata value here should be set to the same as the input_nodata value, otherwise nodata pixels won't be stored correctly